apollo-federation/src/merger/merger.rs

```diff
@@ -1028,8 +1029,124 @@ impl Merger {
+                                &dest.type_name,
+                                "true",
+                                &self.subgraphs[*idx].name,
+                                &dest.type_name,
+                                &self.subgraphs[*idx].name,
+                                human_readable_types(missing_implementations.iter().map(|impl_type| &impl_type.type_name)),
+                                if missing_implementations.len() > 1 { "them" } else { "it" },
+                                &dest.type_name,
+                            )
+                        });
+            }
+        }
+        Ok(has_key)
+    }
+
+    fn validate_interface_objects(
+        &mut self,
+        sources: &Sources<Subgraph<Validated>>,
+        dest: &InterfaceTypeDefinitionPosition,
+    ) -> Result<bool, FederationError> {
+        let supergraph_implementations = self.merged.possible_runtime_types(dest.clone().into())?;
+
+        for (idx, source) in sources.iter() {
+            if source.is_none()
+                || self.subgraphs[*idx].is_interface_object_type(&dest.clone().into())
+            {
+                continue;
+            }
+
+            let subgraph_name = &self.subgraphs[*idx].name;
+            let schema = source.as_ref().unwrap().schema().schema();
+            let defined_implementations: IndexSet<_> = supergraph_implementations
+                .iter()
+                .filter(|implementation| implementation.get(schema).is_ok())
+                .collect::<IndexSet<_>>();
+            if !defined_implementations.is_empty() {
+                self.error_reporter.add_error(CompositionError::InterfaceObjectUsageError {
+                    message: format!("Interface type \"{}\" is defined as an @interfaceObject in subgraph \"{}\" so that subgraph should not define any of the implementation types of \"{}\", but it defines {}",
+                        &dest.type_name,
+                        &subgraph_name,
+                        &dest.type_name,
+                        human_readable_types(defined_implementations.iter().map(|impl_type| &impl_type.type_name)),
+                    )
+                });
+            }
+        }
+
+        Ok(true)
+    }
+
+    pub(crate) fn merge_interface(
+        &mut self,
+        itf: InterfaceTypeDefinitionPosition,
+    ) -> Result<(), FederationError> {
+        let sources: Sources<Subgraph<Validated>> = self
+            .subgraphs
+            .iter()
+            .enumerate()
+            .map(|(idx, subgraph)| (idx, Some(subgraph.clone())))
+            .collect();
+        let has_key = self.validate_interface_keys(&sources, &itf)?;
+
+        let added = self.add_fields_shallow(itf.clone())?;
+
+        for (dest_field, subgraph_fields) in added {
+            if !has_key {
+                let _ = self.hint_on_inconsistent_value_type_field(
```

Comment:
instead of ignoring the return value which is a result we should be propagating the errors with `?` operator